### PR TITLE
cachi2 yarn: purls

### PR DESCRIPTION
**Pull request by slimreaper35**
# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)

**Note:** if the contribution is external (not from an organization member), the CI
pipeline will not run automatically. After verifying that the CI is safe to run:

- [approve GitHub Actions workflows][approve-gh-actions] by clicking a button
- approve the Red Hat Trusted App Pipeline container build by commenting `/ok-to-test`
  (as is the [standard for Pipelines as Code][pac-running-pipeline])

[approve-gh-actions]: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks
[pac-running-pipeline]: https://pipelinesascode.com/docs/guide/running/#running-the-pipelinerun

### CODE REVIEW

**Comment by slimreaper35** [comment]
@everyone PTAL, I have no idea if I am doing it correctly.

**Comment by brunoapimentel** [comment]
One thing that could help the reviews is adding in the PR description a brief example on how each protocol will resolve to a PURL, since the design doc does not get in depth with that. If it's clear to us what is the expected output, then I think the implementation is simpler to code/review.

**Comment by slimreaper35** [comment]
Yeah, but I don't know which attributes from locators are important for PURL generation since the `npm` [specification](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#npm) is very lightweight.

File: cachi2/core/package_managers/yarn/resolver.py
Comment (brunoapimentel) [inline cachi2/core/package_managers/yarn/resolver.py:351]: I believe we can keep the checksums exclusively for HttpsLocator packages. On the other hand, we should report whichever packages didn't have checksums verified, but that's another story (and that will be done via a property, not purls). Right @taylormadore ^^'?

Comment (brunoapimentel) [inline cachi2/core/package_managers/yarn/resolver.py:351]: Another detail we need to handle here is to prefix the checksum with the algorithm used (I'm not sure what it is, we need to look into it). The only place I could find that mentions it in the spec is [here](https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#known-qualifiers-keyvalue-pairs), but there might be more info there.

File: cachi2/core/package_managers/yarn/resolver.py
Comment (brunoapimentel) [inline cachi2/core/package_managers/yarn/resolver.py:370]: We need to take in consideration the `origin` of the repository, otherwise we end up with a local path that won't help locating the package on the internet:

```
pkg:npm/ansi-regex@6.0.1?vcs_url=/home/bpimente/Projects/cachito/test-repos/cachi2-yarn-berry/external-packages/ansi-regex#external-packages/ansi-regex
```

It should look like this:
```
pkg:npm/ansi-regex@6.0.1?vcs_url=git+https://github.com/cachito-testing/cachi2-yarn-berry@70515793108df42547d3320c7ea4cd6b6e505c46#external-packages/ansi-regex
```

File: cachi2/core/package_managers/yarn/resolver.py
Comment (slimreaper35) [reply cachi2/core/package_managers/yarn/resolver.py:351]: [sha512](https://github.com/yarnpkg/berry/blob/017b94ae4eb20dea14ac673a053a1f2974b778ff/packages/yarnpkg-core/sources/hashUtils.ts#L84)

Comment (eskultety) [reply cachi2/core/package_managers/yarn/resolver.py:351]: The other piece of useful information is that they actually use that function for checksums within the lockfile: https://github.com/yarnpkg/berry/blob/017b94ae4eb20dea14ac673a053a1f2974b778ff/packages/yarnpkg-core/sources/Project.ts#L207.
IMO we need to keep references to this information in a code commentary for future reference. This may come handy at some point since the hash algorithm used for checksums hasn't been mentioned anywhere in Yarn docs and thus the algorithm used can be changed by Yarn at will without us even noticing (e.g. when sha512 becomes insecure) ...unless we want to start recomputing checksums  just to make sure the algorithm still checks.

File: cachi2/core/package_managers/yarn/resolver.py
Comment (brunoapimentel) [reply cachi2/core/package_managers/yarn/resolver.py]: Fail it =/
https://github.com/containerbuildsystem/cachi2/blob/main/cachi2/core/scm.py#L49-L56

We should probably revisit this later, as I think it makes for bad user experience.

File: cachi2/core/package_managers/yarn/resolver.py
Comment (taylormadore) [inline cachi2/core/package_managers/yarn/resolver.py:366]: We'll need to pass the path to the root of the git repository here

File: cachi2/core/package_managers/yarn/resolver.py
Comment (brunoapimentel) [inline cachi2/core/package_managers/yarn/resolver.py:354]: The separator between algorithm and checksum should be a `:`. 

https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#generic

File: cachi2/core/package_managers/yarn/resolver.py
Comment (brunoapimentel) [inline cachi2/core/package_managers/yarn/resolver.py:372]: I did a quick test using our test repo as [source](https://github.com/cachito-testing/cachi2-yarn-berry/tree/no-zero-installs). Looking at the results, I got two distinct results for local `strip-ansi:

```sh
# correct, probably derived from the main workspace
pkg:npm/strip-ansi@4.0.0?vcs_url=git%2Bhttps://github.com/cachito-testing/cachi2-yarn-berry%4070515793108df42547d3320c7ea4cd6b6e505c46#external-packages/strip-ansi-4.0.0.tgz

# incorrect, probably derived from packages/then-answer workspace
pkg:npm/strip-ansi@4.0.0?vcs_url=git%2Bhttps://github.com/cachito-testing/cachi2-yarn-berry%4070515793108df42547d3320c7ea4cd6b6e505c46#packages/the-answer/external-packages/strip-ansi-4.0.0.tgz
```

The second one's locator is `file:../../external-packages/strip-ansi-4.0.0.tgz`, so I'm guessing the relative import is not being properly resolved.

**Comment by brunoapimentel** [comment]
Minor nitpick: it feels to me that the `_generate_purl_for_package` function should be a method of `_ComponentResolver`, since it's only ever used by this object, in the context of generation a proper Component.

(I know the change was not introduced in this PR, but maybe we could add a new commit to address this)

File: tests/unit/package_managers/yarn/test_resolver.py
Comment (brunoapimentel) [inline tests/unit/package_managers/yarn/test_resolver.py:959]: I wonder if adding this test here brings any additional value compared to the already existing `test_create_components_single_package` and `test_create_components_patched_packages`, since `_generate_purl_for_package` is called as part of the component creation. Does this bring any additional coverage?

File: tests/unit/package_managers/yarn/test_resolver.py
Comment (brunoapimentel) [inline tests/unit/package_managers/yarn/test_resolver.py:487]: Is it worth adding a scenario that covers packages that live in parent folders? Maybe we missed it the first time because there was no scenario showing it.

**Review by brunoapimentel — COMMENTED** [review]
The implementation LGTM :rocket: 

I only left two comments about tests, but they should be simple enough.

File: tests/unit/package_managers/yarn/test_resolver.py
Comment (slimreaper35) [reply tests/unit/package_managers/yarn/test_resolver.py:959]: No, the coverage is the same. I can leave it there or drop it, does not matter to me. WDYT @everyone

Comment (slimreaper35) [reply tests/unit/package_managers/yarn/test_resolver.py:959]: I will drop it since I move the **_generate_purl_for_package** function to be a method of _**ComponentResolver**

Comment (eskultety) [reply tests/unit/package_managers/yarn/test_resolver.py:959]: > test_create_components_single_package and test_create_components_patched_packages

While the coverage may be the same I wonder if we could potentially decrease the scope of those tests? I'm asking because I think **test_generate_purl_for_package** is the correct test for this implementation, the others  are transitive only which may not be obvious at first glance and one has to go digging what happens underneath.

File: cachi2/core/package_managers/yarn/resolver.py
Comment (eskultety) [inline cachi2/core/package_managers/yarn/resolver.py:381]: Would it be worth it to add a fallthrough `else` branch which would raise an exception about an unknown locator? I know it's highly unlikely a new locator would emerge and we'd have risen an exception in the resolver by now, so I'm just wondering about it, we may as well dismiss that case.

Comment (slimreaper35) [reply cachi2/core/package_managers/yarn/resolver.py]: something like `assert_never` ?

Comment (eskultety) [reply cachi2/core/package_managers/yarn/resolver.py]: Yeah, in fact I think we can just shamelessly copy-paste the whole `else` clause from https://github.com/containerbuildsystem/cachi2/blob/e228b16cfd395b0839348f014b745f1d56d25ef3/cachi2/core/package_managers/yarn/resolver.py#L273

File: tests/unit/package_managers/yarn/test_resolver.py
Comment (brunoapimentel) [reply tests/unit/package_managers/yarn/test_resolver.py]: I think this test's scope is given by the number of different "locator strings vs expected purls" scenarios. I think the currents tests have a good coverage on that.

I do agree that it's less clear to have this specific piece tested as part of `create_components` instead of only `_generate_purl_for_package`, but at the same time, I wonder if it's worth duplicating the scenarios to have a separate test, that's why I raised this point.

**Review by brunoapimentel — APPROVED** [review]
LGTM :rocket:

File: tests/unit/package_managers/yarn/test_resolver.py
Comment (slimreaper35) [reply tests/unit/package_managers/yarn/test_resolver.py:472]: just added the scenario, PTAL

File: cachi2/core/package_managers/yarn/resolver.py
Comment (eskultety) [inline cachi2/core/package_managers/yarn/resolver.py:198]: nitpick: can assign this statement directly in the call below?

File: cachi2/core/package_managers/yarn/resolver.py
Comment (eskultety) [inline cachi2/core/package_managers/yarn/resolver.py:280]: shall we make this method a `staticmethod` since it doesn't interact with `self` in any way?

File: cachi2/core/package_managers/yarn/resolver.py
Comment (eskultety) [inline cachi2/core/package_managers/yarn/resolver.py:338]: This commit should have happened before the first one, i.e. before implementing the logic, that way we'd end up with an overall smaller change diff.

File: cachi2/core/package_managers/yarn/resolver.py
Comment (eskultety) [inline cachi2/core/package_managers/yarn/resolver.py:290]: This change should have been part of the commit which implements the guts of the method: https://github.com/containerbuildsystem/cachi2/pull/374/commits/0709ff1cdb317fbcf363637028842beca45e1341

**Comment by eskultety** [comment]
Regarding functionality, I don't have any comments as @brunoapimentel has done a good job at digging into the details.

File: cachi2/core/package_managers/yarn/resolver.py
Comment (slimreaper35) [reply cachi2/core/package_managers/yarn/resolver.py:338]: Squashed first and last two commits into one commit, which implements a static method of `_ComponentResolver` class for generating PURLs. Everyone okay with that before I force push :fearful: ?

File: cachi2/core/package_managers/yarn/resolver.py
Comment (eskultety) [reply cachi2/core/package_managers/yarn/resolver.py:290]: I meant the whole hunk, not just the single line ;) .

File: cachi2/core/package_managers/yarn/resolver.py
Comment (eskultety) [reply cachi2/core/package_managers/yarn/resolver.py:338]: just make sure unit tests are passing for each commit in the PR (like they are now).

File: cachi2/core/package_managers/yarn/resolver.py
Comment (eskultety) [reply cachi2/core/package_managers/yarn/resolver.py:198]: @slimreaper35 you pushed a new revision, but didn't respond to this comment. I'm okay if we ignore this, but a resolution comment would be nice :).

File: cachi2/core/package_managers/yarn/resolver.py
Comment (eskultety) [reply cachi2/core/package_managers/yarn/resolver.py:338]: @slimreaper35 after thoroughly examining the latest revision what I meant by

> This commit should have happened before the first one, i.e. before implementing the logic, that way we'd end up with an overall smaller change diff.

was that:

- in commit 1 you convert `generate_purl_for_package` to a static instance method
- in commit 2 you provide the actual implementation for the method
- in commit 3 you take care of the unit tests

If you squash everything in a single commit that (from logic POV) adds too many changes in a single commit whereas we want gradual well-structured changes.

I apologize for the misunderstanding that may have originated from my comment

> just make sure unit tests are passing for each commit in the PR (like they are now).

as I only put ^this comment in not to forget about running unit tests after each patch without actually paying much attention to the details you posted before me leading to the confusion that I may have approved of squashing the 2 commits together. After all, it's better to just push a new revision right away and see if there are more comments rather than waiting for approval for future changes in the next revision.

File: cachi2/core/package_managers/yarn/resolver.py
Comment (slimreaper35) [reply cachi2/core/package_managers/yarn/resolver.py:198]: It was just my personal visual preference of putting the assignment of purl on a separate line. There is no other reason behind it, so, I am okay with both options as well.

File: cachi2/core/package_managers/yarn/resolver.py
Comment (eskultety) [reply cachi2/core/package_managers/yarn/resolver.py:338]: @slimreaper35 do you plan on taking care of the commits as suggested ^above?

Comment (slimreaper35) [reply cachi2/core/package_managers/yarn/resolver.py:338]: I will try

**Review by eskultety — APPROVED** [review]

**Pull request merged**
